### PR TITLE
[RPC tests] Fix @_skip_if_tensorpipe always skipping for all agents

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40823 [RPC tests] Enroll TensorPipe in missing test suites
* #40822 [RPC tests] Remove global TEST_CONFIG
* #40821 [RPC tests] Move some functions to methods of fixture
* #40820 [RPC tests] Make generic fixture an abstract base class
* #40819 [RPC tests] Avoid decorators to skip tests
* #40818 [RPC tests] Merge process group tests into single entry point
* #40817 [RPC tests] Merge tests for faulty agent into single script
* #40816 [RPC tests] Merge TensorPipe tests into single entry point
* #40913 [RPC tests] Fix file descriptor leak
* #40815 [RPC tests] Align ddp_under_dist_autograd test with others
* #40814 [RPC tests] Remove world_size and init_method from TensorPipe fixture
* **#40860 [RPC tests] Fix @_skip_if_tensorpipe always skipping for all agents**

It turns out that the `@_skip_if_tensorpipe_agent` decorator was written in such a way that it accidentally caused the test to become a no-op (and thus always succeed) for all agents. What this means is that all tests wrapped by that decorator were never ever being run, for any agent.

My understanding of the root cause is that the following code:
```
@_skip_if_tensorpipe_agent
def test_foo(self):
    self.assertEqual(2 + 2, 4)
```
ended up behaving somewhat like this:
```
def test_foo(self):
    def original_test_func(self):
        self.assertEqual(2 + 2, 4)
    return unittest.skipIf(self.agent == "TENSORPIPE")(original_test_func)
```
which means that the test body of the decorated method was not actually calling the original test method.

This issue probably came from the `@_skip_if_tensorpipe_agent` being copy-pasted from `@requires_process_group_agent` (which, however, is not a decorator but rather a decorator *factory*). An unfortunate naming (calling `decorator` what was in fact the wrapped method) then hindered readability and hid the issue.

Note that a couple of tests had become legitimately broken in the meantime and no one had noticed. The breakages have been introduced in #39909 (a.k.a., D22011868).

Differential Revision: [D22332611](https://our.internmc.facebook.com/intern/diff/D22332611/)

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 1]
```diff
@@ -472,14 +472,18 @@ def return_future():
...
-            )(old_func)
+    def _skip_if_tensorpipe_agent(test_func):  # noqa: B902
+        @wraps(test_func)
+        def wrapped_test_func(self):
+            if self.rpc_backend == rpc.backend_registry.BackendType.TENSORPIPE:
+                # In theory we should call self.skipTest(reason), but this does
+                # not play well with the test runner, which sees an exception
+                # being raised and marks the test as failed. Thus instead we
+                # make the test a successful no-op.
+                return
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:483]: Do we need to call `sys.exit()` and provide an skip code here?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 2]
```diff
@@ -472,14 +472,18 @@ def return_future():
...
+            test_func(self)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:484]: should this be `return test_func(self)` in case the function returns anything?

**Review by mrshenli — APPROVED** [review]
Thanks for fixing!

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 1 above)
Comment (lw) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:483]: Do you mean these?
https://github.com/pytorch/pytorch/blob/1a74bb84f2b0fd0e1a532941645e24e2cfea1e6a/torch/testing/_internal/common_distributed.py#L26-L33

I wasn't aware of these. We could in fact add a new code just for those. Note however that this decorator will go away later in the stack (in #40819). So I don't know if it makes sense to invest that much into it now just to then undo it all.

Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:483]: Oh I see. Haven't reached #40819 yet. In that case, please ignore this comment.

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 2 above)
Comment (lw) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:484]: I was importing the accept into the internal diff and this triggered a land which didn't pick up this fix (as it was still a draft). If that's a problem I can send out a PR that fixes it. LMK.

**Pull request closed without merging**